### DISCUSSION TITLE
A listing of files generated from a template

### DISCUSSION THREAD

**Discussion by mbojan** [quarto-dev/quarto-cli#961] [Show and Tell]
Hi! I'm came back to playing with Quarto and the usecase of generating a website containing a collection of reports for subsets of some data. I've made a toy example at https://mbojan.github.io/try-quarto (the code is on the [`website-cars`](https://github.com/mbojan/try-quarto/tree/website-cars) branch). There the data is `mtcars` and the reports correspond to its rows (cars), but the general idea should be applicable to any subsettable R object (grouped data frame, a list of some model objects etc.). I'm exploring two features:

- Using Quarto's listing to display a list/grid of individual reports (ideally the workflow should be agnostic to the number of data elements, here rows, to generate the reports for). 
- Generate the reports from the common template (e.g. parametrised `qmd` file) with a data-generated content, but also metadata (title, description, etc.).

I came up with a setup that has the report template in `whisker` format, which is used in a pre-render script to populate folder `cars` with `qmd` files for each car. These `qmd`s are then rendered by Quarto normally.

Some questions/observations:

1. Is there a natural way to have such report template as a `qmd` file (and abandon `whisker`) which is repeatedly rendered by Quarto but with different parameters/metadata? I think I tried executing `quarto` CLI by a pre-render script, but as the whole thing is a Quarto Project, it went astray and tried to render other files too.
2. (Ad second bullet above) At this moment, unlike RMarkdown, I can't use within-document code to generate YAML metadata (c.f. #606 ). To use Quarto to render the reports from the template I'd need to pass both the parameters (so that they can be handled by code) and the metadata (to set YAML fields properly) at the same time. Correct?
3. Rendering the whole thing takes quite a bit of time even though individual reports are computationally trivial. I'm wondering how to speed things up. Ideally the reports could be processed in parallel. Is it safe to execute several `quarto render somefile.qmd` in parallel within a Quarto Project or would they get in each others' way by writing to `.quarto` folder?
4. Is there any other Quarto functionality that one could leverage in this usecase?

Thanks!

**Comment by jjallaire** [comment]
Currently there is no way to parameterized individual `.qmd` file renders within a project (rather you need to generate the entire project dynamically as you are doing). 

It's correct that you can't dynamically generate the metadata (as we need to compute on it at various junctures before the `.qmd` is rendered).

It's definitely not safe to execute multiple quarto render invocations concurrently on the same `.qmd`.

As you've discovered the built-in project mechanism doesn't handle this all that well. I can definitely forsee us making improvements here but in the meantime "going meta" and dynamically generating the project as you are doing is the way to go.

**Reply by mbojan** [reply]
Thanks @jjallaire for these remarks and in fact for the whole tool because I find (1) working with it a lot cleaner than pure RMarkdown and (2) the templates are much nicer. As to:

> It's correct that you can't dynamically generate the metadata (as we need to compute on it at various junctures before the `.qmd` is rendered).

Right. Once "one goes meta" then passing metadata (e.g. `-M`) or using shortcodes for variables, either environment or through `_variables.yml` it is possible too, although perhaps not too elegant/straightforward (thx @cderv for https://github.com/quarto-dev/quarto-cli/discussions/606#discussioncomment-2550278).

> As you've discovered the built-in project mechanism doesn't handle this all that well. I can definitely forsee us making improvements here but in the meantime "going meta" and dynamically generating the project as you are doing is the way to go.

Sounds good :)

**Comment by VirginiaWoloshen** [comment]
I would also like to do this. Currently, I've been using a parameterized quarto template to generate reports. I pass the parameter to a quarto_render(input = "template.qmd", 
              execute_params = list(param1 = param_example), 
              output_file = paste("./", param_example, "_report.html", sep = ""))
but this renders the html file which is fine for one off events but I would like to auto-generate a blog post on a Quarto blog with this. Can the output_file be an .qmd file (required for blog post listings)?

**Reply by mcanouil** [reply]
Probably better to open a new discussion in Q&A, it does not feel your question is quite the same.

Anyway, it seems child code-cell generation might help: https://github.com/quarto-dev/quarto-cli/discussions/2094

**Comment by gavin-kelly-1** [comment]
I've produced a makefile-based solution to this.  Briefly, every 'to-be-parametrised' qmd file has a partner quasi-yaml file  (which contains multiple metadata fields and parameter sets, so e.g. `title` and `params` appear multiple times within the same file, so it's not truly valid yaml per se), which then gets pre-processed into a staging directory: each 'parent' qmd gets iterated across the multiple parameter-sets, and the nth chunk from the yml gets injected into the nth instance of that qmd.  That staging directory can then be rendered to produce a website of inter-related pages.

I've put a repository at https://github.com/FrancisCrickInstitute/babs-pre-quarto if anyone's interested.  It's working for me at least, and even if it doesn't exactly work as you need it, might offer some ideas.
